// whitestormjs/lib/core/Camera.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Camera = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _get2 = require('babel-runtime/helpers/get');

var _get3 = _interopRequireDefault(_get2);

var _three = require('three');

var THREE = _interopRequireWildcard(_three);

var _Loop = require('../extras/Loop');

var _CoreObject2 = require('./CoreObject');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Camera = function (_CoreObject) {
  (0, _inherits3.default)(Camera, _CoreObject);

  function Camera(params) {
    var type = arguments.length <= 1 || arguments[1] === undefined ? 'camera' : arguments[1];

    var _ret;

    var localWindow = arguments.length <= 2 || arguments[2] === undefined ? window : arguments[2];
    (0, _classCallCheck3.default)(this, Camera);

    if (!type) console.error('@constructor: Please specify " type ".');

    var _set = function _set(x, y, z) {
      _this.x = x;
      _this.y = y;
      _this.z = z;
    };

    params.useTarget = Boolean(params.target);

    var _this = (0, _possibleConstructorReturn3.default)(this, (Camera.__proto__ || Object.getPrototypeOf(Camera)).call(this, {
      camera: {
        fov: 45,
        aspect: localWindow.innerWidth / localWindow.innerHeight,
        near: 1,
        far: 1000,
        left: localWindow.innerWidth / -2,
        right: localWindow.innerWidth / 2,
        top: localWindow.innerHeight / 2,
        bottom: localWindow.innerHeight / -2,
        cubeResolution: 128
      },

      helper: false,

      pos: {
        x: 0,
        y: 0,
        z: 0,
        set: _set
      },

      rot: {
        x: 0,
        y: 0,
        z: 0,
        set: _set
      },

      target: {
        x: 0,
        y: 0,
        z: 0,
        set: _set
      }
    }));

    if (params instanceof THREE.Camera) {
      (0, _get3.default)(Camera.prototype.__proto__ || Object.getPrototypeOf(Camera.prototype), 'setParams', _this).call(_this, {
        pos: { x: params.position.x, y: params.position.y, z: params.position.z },
        rot: { x: params.rotation.x, y: params.rotation.y, z: params.rotation.z }
      });
    } else (0, _get3.default)(Camera.prototype.__proto__ || Object.getPrototypeOf(Camera.prototype), 'setParams', _this).call(_this, params);

    var scope = Object.assign(_this, {
      type: type,
      helper: false
    });

    if (params instanceof THREE.Camera) _this.setNative(params);

    return _ret = scope, (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(Camera, [{
    key: 'wrap',
    value: function wrap() {
      var _this2 = this;

      for (var _len = arguments.length, tags = Array(_len), _key = 0; _key < _len; _key++) {
        tags[_key] = arguments[_key];
      }

      return new Promise(function (resolve, reject) {
        var _native = _this2.getNative(),
            _params = _this2.getParams();

        _this2.position.set(_params.pos.x, _params.pos.y, _params.pos.z);

        _this2.rotation.set(_params.rot.x, _params.rot.y, _params.rot.z);

        if (_params.useTarget) _this2.lookAt(_params.target);

        if (_params.helper) _this2.helper = new THREE.CameraHelper(_native);

        tags.forEach(function (tag) {
          _this2[tag] = true;
        });

        _this2.emit('ready');

        resolve(_this2);
      });
    }
  }, {
    key: 'addTo',
    value: function addTo(parent) {
      this.parent = parent;

      var _helper = this.helper,
          _scope = this;

      return new Promise(function (resolve, reject) {
        try {
          _scope.parent.getScene().add(_scope.getNative());
          _scope.parent.children.push(_scope);

          if (_helper) _scope.parent.getScene().add(_helper);
        } catch (err) {
          console.error(err.message);
          reject();
        } finally {
          resolve(_scope);
          _scope.emit('ready');
        }
      });
    }
  }, {
    key: 'clone',
    value: function clone() {
      return new Camera(this.getParams(), this.type).copy(this);
    }
  }, {
    key: 'copy',
    value: function copy(source) {
      if (source.getNative()) {
        this.setNative(source.getNative().clone());
        this.setParams(source.getParams());

        this.wrap();

        this.position = source.position.clone();
        this.rotation = source.rotation.clone();
      } else this.setParams(source.getParams());

      this.type = source.type;

      return this;
    }
  }, {
    key: 'getParent',
    value: function getParent() {
      return this.parent;
    }
  }, {
    key: 'follow',
    value: function follow(curve) {
      var time = arguments.length <= 1 || arguments[1] === undefined ? 1000 : arguments[1];
      var loop = arguments[2];
      var lookAt = arguments[3];

      var _scope = this,
          gEnd = time;

      var animation = new _Loop.Loop(function (clock) {
        var u = clock.getElapsedTime() * 1000 / gEnd,
            vec1 = curve.getPoint(u),
            vec2 = curve.getPoint((u + 0.01) % 1);

        _scope.position.set(vec1.x, vec1.y, vec1.z);

        if (!lookAt) _scope.lookAt(vec2);else if (lookAt instanceof TVector3) _scope.lookAt(lookAt);else if (lookAt instanceof TCurve || lookAt instanceof TCurvePath) _scope.lookAt(lookAt.getPoint(u));
      });

      animation.start();

      if (loop) {
        setInterval(function () {
          animation.stop();

          animation = new _Loop.Loop(function (clock) {
            var u = clock.getElapsedTime() * 1000 / gEnd,
                vec1 = curve.getPoint(u),
                vec2 = curve.getPoint((u + 0.01) % 1);

            _scope.position.set(vec1.x, vec1.y, vec1.z);

            if (!lookAt) _scope.lookAt(vec2);else if (lookAt instanceof TVector3) _scope.lookAt(lookAt);else if (lookAt instanceof TCurve || lookAt instanceof TCurvePath) _scope.lookAt(lookAt.getPoint(u));
          });

          animation.start();
        }, time);
      } else {
        setTimeout(function () {
          animation.stop();
        }, time);
      }
    }
  }, {
    key: 'lookAt',
    value: function lookAt(vector3) {
      return this.getNative().lookAt(vector3);
    }
  }, {
    key: 'getWorldDirection',
    value: function getWorldDirection(vector3) {
      return this.getNative().getWorldDirection(vector3);
    }
  }, {
    key: 'position',
    get: function get() {
      return this.getNative().position;
    },
    set: function set(vector3) {
      return this.getNative().position.copy(vector3);
    }
  }, {
    key: 'rotation',
    get: function get() {
      return this.getNative().rotation;
    },
    set: function set(euler) {
      return this.getNative().rotation.copy(euler);
    }
  }, {
    key: 'quaternion',
    get: function get() {
      return this.getNative().quaternion;
    },
    set: function set(euler) {
      return this.getNative().quaternion.copy(euler);
    }
  }]);
  return Camera;
}(_CoreObject2.CoreObject);

exports.Camera = Camera;
