// whitestormjs/lib/core/CoreObject.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CoreObject = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _minivents = require('minivents');

var _minivents2 = _interopRequireDefault(_minivents);

var _index = require('../utils/index');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CoreObject = function () {
  function CoreObject() {
    var defaults = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
    var structurable = arguments.length <= 1 || arguments[1] === undefined ? true : arguments[1];
    (0, _classCallCheck3.default)(this, CoreObject);

    var scope = structurable ? Object.assign(this, {
      __whsobject: true,
      __releaseTime: new Date().getTime(),
      __params: {},
      __defaults: defaults,

      parent: null,
      children: []
    }, new _minivents2.default()) : Object.assign(this, {
      __whsobject: true,
      __releaseTime: new Date().getTime(),
      __params: {},
      __defaults: defaults
    }, new _minivents2.default());

    return scope;
  }

  (0, _createClass3.default)(CoreObject, [{
    key: 'setParams',
    value: function setParams() {
      var params = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

      this.__params = (0, _index.extend)(params, this.__defaults);
      return this.__params;
    }
  }, {
    key: 'updateParams',
    value: function updateParams() {
      var params = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

      this.__params = (0, _index.extend)(params, this.__params);
      return this.__params;
    }
  }, {
    key: 'getParams',
    value: function getParams() {
      return this.__params;
    }
  }, {
    key: 'setNative',
    value: function setNative(native) {
      this._native = native;
      return this.native;
    }
  }, {
    key: 'getNative',
    value: function getNative() {
      return this._native;
    }
  }, {
    key: 'add',
    value: function add(children) {
      var _scope = this;

      if (children.addTo) return children.addTo(this);else if (children instanceof Object) {
        return new Promise(function (resolve) {
          children.parent = _scope;

          _scope.getNative().add(children.getNative());
          _scope.children.push(_scope);

          resolve();
        });
      }
    }
  }, {
    key: 'remove',
    value: function remove(source) {
      this.getNative().remove(source.getNative());

      this.children.splice(this.children.indexOf(source), 1);
      source.parent = null;
      source.emit('remove');

      return this;
    }
  }]);
  return CoreObject;
}();

exports.CoreObject = CoreObject;
