// whitestormjs/lib/core/Light.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Light = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _get2 = require('babel-runtime/helpers/get');

var _get3 = _interopRequireDefault(_get2);

var _three = require('three');

var THREE = _interopRequireWildcard(_three);

var _Loop = require('../extras/Loop');

var _World = require('./World');

var _CoreObject2 = require('./CoreObject');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Light = function (_CoreObject) {
  (0, _inherits3.default)(Light, _CoreObject);

  function Light(params) {
    var _ret;

    var type = arguments.length <= 1 || arguments[1] === undefined ? 'light' : arguments[1];
    (0, _classCallCheck3.default)(this, Light);

    var _set = function _set(x, y, z) {
      _this.x = x;
      _this.y = y;
      _this.z = z;
    };

    var _this = (0, _possibleConstructorReturn3.default)(this, (Light.__proto__ || Object.getPrototypeOf(Light)).call(this, {
      light: {
        color: 0xffffff,
        skyColor: 0xffffff,
        groundColor: 0xffffff,

        intensity: 1,
        distance: 100,
        angle: Math.PI / 3,
        exponent: 0,
        decay: 1
      },

      helper: false,

      shadowmap: {
        cast: true,

        bias: 0,
        radius: 1,

        width: 1024,
        height: 1024,

        near: true,
        far: 400,
        fov: 60,

        top: 200,
        bottom: -200,
        left: -200,
        right: 200
      },

      pos: {
        x: 0,
        y: 0,
        z: 0,
        set: _set
      },

      rot: {
        x: 0,
        y: 0,
        z: 0,
        set: _set
      },

      target: {
        x: 0,
        y: 0,
        z: 0,
        set: _set
      }
    }));

    if (params instanceof THREE.Light) {
      (0, _get3.default)(Light.prototype.__proto__ || Object.getPrototypeOf(Light.prototype), 'setParams', _this).call(_this, {
        pos: { x: params.position.x, y: params.position.y, z: params.position.z },
        rot: { x: params.rotation.x, y: params.rotation.y, z: params.rotation.z }
      });
    } else (0, _get3.default)(Light.prototype.__proto__ || Object.getPrototypeOf(Light.prototype), 'setParams', _this).call(_this, params);

    _this.type = type;

    if (params instanceof THREE.Light) _this.setNative(params);

    return _ret = _this, (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  /**
   * Applying shadow & position & rotation.
   *
   * @param {...String} tags - Tags that defines what to do with light
   * additionally.
   */


  (0, _createClass3.default)(Light, [{
    key: 'wrap',
    value: function wrap() {
      var _this2 = this;

      for (var _len = arguments.length, tags = Array(_len), _key = 0; _key < _len; _key++) {
        tags[_key] = arguments[_key];
      }

      return new Promise(function (resolve, reject) {
        var _native = _this2.getNative(),
            _params = _this2.getParams();

        if (tags.indexOf('no-shadows') < 0) {
          _this2.wrapShadow();
        }

        if (tags.indexOf('no-transforms') < 0) {
          _this2.position.set(_params.pos.x, _params.pos.y, _params.pos.z);

          _this2.rotation.set(_params.rot.x, _params.rot.y, _params.rot.z);

          if (_this2.target) _this2.target = _params.target;
        }

        tags.forEach(function (tag) {
          _this2[tag] = true;
        });

        _this2.emit('ready');

        resolve(_this2);
      });
    }
  }, {
    key: 'addTo',
    value: function addTo(parent) {
      var _this3 = this;

      this.parent = parent;

      return new Promise(function (resolve, reject) {
        var _native = _this3.getNative(),
            _parent = _this3.parent;

        var parentNative = _parent instanceof _World.World ? _parent.getScene() : _parent.getNative();

        parentNative.add(_native);
        parent.children.push(_this3);

        if (_this3.helper) parentNative.add(_this3.helper);
        if (_native.target) parentNative.add(_native.target);

        resolve(_this3);
        _this3.emit('ready');
      });
    }

    /**
     * Set shadow properties for light.
     */

  }, {
    key: 'wrapShadow',
    value: function wrapShadow() {
      var _this4 = this;

      return new Promise(function (resolve, reject) {
        var _native = _this4.getNative(),
            _shadow = _this4.getParams().shadowmap;

        _native.castShadow = _shadow.cast;
        _native.shadow.mapSize.width = _shadow.width;
        _native.shadow.mapSize.height = _shadow.height;
        _native.shadow.bias = _shadow.bias;
        _native.shadow.radius = _shadow.radius;

        var _shadowCamera = _native.shadow.camera;

        _shadowCamera.near = _shadow.near;
        _shadowCamera.far = _shadow.far;
        _shadowCamera.fov = _shadow.fov;

        _shadowCamera.left = _shadow.left;
        _shadowCamera.right = _shadow.right;
        _shadowCamera.top = _shadow.top;
        _shadowCamera.bottom = _shadow.bottom;

        resolve(_this4);
      });
    }
  }, {
    key: 'clone',
    value: function clone() {
      return new Light(this.getParams(), this.type).copy(this);
    }
  }, {
    key: 'copy',
    value: function copy(source) {
      if (source.getNative()) {
        this.setNative(source.getNative().clone());
        if (source.helper) this.helper = source.helper.clone();
        this.setParams(source.getParams());

        this.wrap();

        this.position = source.position.clone();
        this.rotation = source.rotation.clone();
        if (source.target) this.target = source.target.clone();
      } else this.setParams(source.getParams());

      this.type = source.type;

      return this;
    }
  }, {
    key: 'getParent',
    value: function getParent() {
      return this.parent;
    }
  }, {
    key: 'getWorld',
    value: function getWorld() {
      var p = this.parent;

      while (!(p instanceof _World.World)) {
        if (p) p = p.parent;else return false;
      }

      return p;
    }
  }, {
    key: 'follow',
    value: function follow(curve) {
      var time = arguments.length <= 1 || arguments[1] === undefined ? 1000 : arguments[1];
      var loop = arguments[2];
      var lookAt = arguments[3];

      var _scope = this,
          gEnd = time;

      var animation = new _Loop.Loop(function (clock) {
        var u = clock.getElapsedTime() * 1000 / gEnd,
            vec1 = curve.getPoint(u),
            vec2 = curve.getPoint((u + 0.01) % 1);

        _scope.position.set(vec1.x, vec1.y, vec1.z);

        if (!lookAt) _scope.lookAt(vec2);else if (lookAt instanceof THREE.Vector3) _scope.lookAt(lookAt);else if (lookAt instanceof THREE.Curve || lookAt instanceof THREE.CurvePath) _scope.lookAt(lookAt.getPoint(u));
      });

      animation.start();

      if (loop) {
        setInterval(function () {
          animation.stop();

          animation = new _Loop.Loop(function (clock) {
            var u = clock.getElapsedtime() * 1000 / gEnd,
                vec1 = curve.getPoint(u),
                vec2 = curve.getPoint((u + 0.01) % 1);

            _scope.position.set(vec1.x, vec1.y, vec1.z);

            if (!lookAt) _scope.lookAt(vec2);else if (lookAt instanceof THREE.Vector3) _scope.lookAt(lookAt);else if (lookAt instanceof THREE.Curve || lookAt instanceof THREE.CurvePath) _scope.lookAt(lookAt.getPoint(u));
          });

          animation.start();
        }, time);
      } else {
        setTimeout(function () {
          animation.stop();
        }, time);
      }
    }

    /* VISIBILITY */

  }, {
    key: 'show',
    value: function show() {
      this.getNative().visible = true;
    }
  }, {
    key: 'hide',
    value: function hide() {
      this.getNative().visible = false;
    }
  }, {
    key: 'position',
    get: function get() {
      return this.getNative().position;
    },
    set: function set(vector3) {
      return this.getNative().position.copy(vector3);
    }
  }, {
    key: 'rotation',
    get: function get() {
      return this.getNative().rotation;
    },
    set: function set(euler) {
      return this.getNative().rotation.copy(euler);
    }
  }, {
    key: 'quaternion',
    get: function get() {
      return this.getNative().quaternion;
    },
    set: function set(euler) {
      return this.getNative().quaternion.copy(euler);
    }
  }, {
    key: 'target',
    get: function get() {
      return this.getNative().target;
    },
    set: function set(vector3) {
      if (vector3 instanceof THREE.Object3D) this.getNative().target.copy(vector3); // THREE.Object3D in this case.
      else this.getNative().target.position.copy(vector3);
    }
  }]);
  return Light;
}(_CoreObject2.CoreObject);

exports.Light = Light;
