// whs/lib/extras/Loop.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Loop = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _three = require('three');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Loop = function () {
  function Loop(func) {
    (0, _classCallCheck3.default)(this, Loop);

    this.func = func;
    this.clock = new _three.Clock();
    this.enabled = false;
  }

  (0, _createClass3.default)(Loop, [{
    key: 'start',
    value: function start(world) {
      if (world) world.addLoop(this);

      this.clock.start();
      this.enabled = true;
    }
  }, {
    key: 'stop',
    value: function stop(world) {
      this.clock.stop();
      this.enabled = false;

      if (world) world.removeLoop(this);
    }
  }, {
    key: 'execute',
    value: function execute() {
      return this.func(this.clock);
    }
  }]);
  return Loop;
}();

exports.Loop = Loop;
