// whs/lib/extras/VirtualMouse.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VirtualMouse = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _three = require('three');

var _minivents = require('minivents');

var _minivents2 = _interopRequireDefault(_minivents);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var VirtualMouse = exports.VirtualMouse = function (_Events) {
  (0, _inherits3.default)(VirtualMouse, _Events);

  function VirtualMouse(world) {
    (0, _classCallCheck3.default)(this, VirtualMouse);

    var _this = (0, _possibleConstructorReturn3.default)(this, (VirtualMouse.__proto__ || (0, _getPrototypeOf2.default)(VirtualMouse)).call(this));

    _this.mouse = new _three.Vector2();
    _this.raycaster = new _three.Raycaster();
    _this.world = null;
    _this.canvas = null;
    _this.projectionPlane = new _three.Plane(new _three.Vector3(0, 0, 1), 0);


    world.mouse = _this;
    _this.world = world;

    if (world.renderer) _this.canvas = world.renderer.domElement;

    world.$element.addEventListener('mousemove', _this.update.bind(_this));
    world.$element.addEventListener('click', function () {
      return _this.emit('click');
    });
    world.$element.addEventListener('mousedown', function () {
      return _this.emit('mousedown');
    });
    world.$element.addEventListener('mouseup', function () {
      return _this.emit('mouseup');
    });
    return _this;
  }

  (0, _createClass3.default)(VirtualMouse, [{
    key: 'update',
    value: function update(e) {
      var rect = this.canvas.getBoundingClientRect();

      this.mouse.x = (e.clientX - rect.left) / (rect.right - rect.left) * 2 - 1;
      this.mouse.y = -((e.clientY - rect.top) / (rect.bottom - rect.top)) * 2 + 1;

      this.projectionPlane.normal.copy(this.world.$camera.native.getWorldDirection());

      this.raycaster.setFromCamera(this.mouse, this.world.$camera.native);
      this.emit('move');
    }
  }, {
    key: 'track',
    value: function track(component) {
      var _this2 = this;

      var isHovered = false;

      this.on('move', function () {
        if (_this2.hovers(component)) {
          if (isHovered) component.emit('mousemove');else {
            component.emit('mouseover');
            isHovered = true;
          }
        } else if (isHovered) {
          component.emit('mouseout');
          isHovered = false;
        }
      });

      this.on('click', function () {
        if (isHovered) component.emit('click');else component.emit('offClick');
      });

      this.on('mousedown', function () {
        if (isHovered) component.emit('mousedown');
      });

      this.on('mouseup', function () {
        if (isHovered) component.emit('mouseup');
      });
    }
  }, {
    key: 'intersection',
    value: function intersection(component) {
      return this.raycaster.intersectObject(component.native);
    }
  }, {
    key: 'project',
    value: function project() {
      var plane = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.projectionPlane;

      return this.raycaster.ray.intersectPlane(plane);
    }
  }, {
    key: 'hovers',
    value: function hovers(component) {
      var intersection = this.intersection(component)[0];
      return intersection ? intersection.object === component.native : false;
    }
  }, {
    key: 'ray',
    get: function get() {
      return this.raycaster.ray;
    }
  }, {
    key: 'x',
    get: function get() {
      return this.mouse.x;
    }
  }, {
    key: 'y',
    get: function get() {
      return this.mouse.y;
    }
  }]);
  return VirtualMouse;
}(_minivents2.default);
