// whs/lib/components/meshes/Cylinder.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Cylinder = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _get2 = require('babel-runtime/helpers/get');

var _get3 = _interopRequireDefault(_get2);

var _class, _class2, _temp;

var _three = require('three');

var _index = require('../../physics/index.js');

var _Component2 = require('../../core/Component');

var _MeshComponent = require('../../core/MeshComponent');

var _PhysicsComponent = require('../../core/PhysicsComponent');

var _SoftbodyComponent = require('../../core/SoftbodyComponent');

var _index2 = require('../../utils/index');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Cylinder = (0, _SoftbodyComponent.SoftbodyComponent)(_class = (0, _PhysicsComponent.PhysicsComponent)(_class = (0, _MeshComponent.MeshComponent)(_class = (_temp = _class2 = function (_Component) {
  (0, _inherits3.default)(Cylinder, _Component);

  function Cylinder() {
    var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    (0, _classCallCheck3.default)(this, Cylinder);

    var _this = (0, _possibleConstructorReturn3.default)(this, (Cylinder.__proto__ || (0, _getPrototypeOf2.default)(Cylinder)).call(this, params, Cylinder.defaults, Cylinder.instructions));

    if (params.build) {
      _this.build(params);
      (0, _get3.default)(Cylinder.prototype.__proto__ || (0, _getPrototypeOf2.default)(Cylinder.prototype), 'wrap', _this).call(_this);
    }
    return _this;
  }

  (0, _createClass3.default)(Cylinder, [{
    key: 'build',
    value: function build() {
      var _this2 = this;

      var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      var material = (0, _index2.loadMaterial)(params.material);

      var MeshNative = void 0;

      if (this.physics && this.params.softbody) MeshNative = _index.SoftMesh;else if (this.physics) MeshNative = _index.CylinderMesh;else MeshNative = _three.Mesh;

      return new _promise2.default(function (resolve) {
        _this2.native = new MeshNative(_this2.buildGeometry(params), material, _this2.params);

        resolve();
      });
    }
  }, {
    key: 'buildGeometry',
    value: function buildGeometry() {
      var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      var GConstruct = params.buffer && !params.softbody ? _three.CylinderBufferGeometry : _three.CylinderGeometry;

      var geometry = new GConstruct(params.geometry.radiusTop, params.geometry.radiusBottom, params.geometry.height, params.geometry.radiusSegments, params.geometry.heightSegments, params.geometry.openEnded, params.geometry.thetaStart, params.geometry.thetaLength);

      if (params.softbody) this.proccessSoftbodyGeometry(geometry);

      return geometry;
    }
  }, {
    key: 'clone',
    value: function clone() {
      return new Cylinder({ build: false }).copy(this);
    }
  }, {
    key: 'g_radiusTop',
    set: function set(val) {
      this._native.geometry = this.buildGeometry(this.updateParams({ geometry: { radiusTop: val } }));
    },
    get: function get() {
      return this._native.geometry.parameters.radiusTop;
    }
  }, {
    key: 'g_radiusBottom',
    set: function set(val) {
      this._native.geometry = this.buildGeometry(this.updateParams({ geometry: { radiusBottom: val } }));
    },
    get: function get() {
      return this._native.geometry.parameters.radiusBottom;
    }
  }, {
    key: 'g_height',
    set: function set(val) {
      this._native.geometry = this.buildGeometry(this.updateParams({ geometry: { height: val } }));
    },
    get: function get() {
      return this._native.geometry.parameters.height;
    }
  }, {
    key: 'g_radiusSegments',
    set: function set(val) {
      this._native.geometry = this.buildGeometry(this.updateParams({ geometry: { radiusSegments: val } }));
    },
    get: function get() {
      return this._native.geometry.parameters.radiusSegments;
    }
  }]);
  return Cylinder;
}(_Component2.Component), _class2.defaults = (0, _extends3.default)({}, _Component2.Component.defaults, {
  geometry: {
    radiusTop: 0,
    radiusBottom: 1,
    height: 1,
    radiusSegments: 32,
    heightSegments: 1,
    openEnded: false,
    thetaStart: 0,
    thetaLength: Math.PI * 2
  }
}), _class2.instructions = (0, _extends3.default)({}, _Component2.Component.instructions, {
  geometry: ['radiusTop', 'radiusBottom', 'height', 'radiusSegments', 'heightSegments', 'openEnded', 'thetaStart', 'thetaLength']
}), _temp)) || _class) || _class) || _class;

exports.Cylinder = Cylinder;
