// whs/lib/components/meshes/Morph.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Morph = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _get2 = require('babel-runtime/helpers/get');

var _get3 = _interopRequireDefault(_get2);

var _class, _class2, _temp;

var _three = require('three');

var _index = require('../../physics/index.js');

var Physijs = _interopRequireWildcard(_index);

var _Component2 = require('../../core/Component');

var _MeshComponent = require('../../core/MeshComponent');

var _PhysicsComponent = require('../../core/PhysicsComponent');

var _SoftbodyComponent = require('../../core/SoftbodyComponent');

var _index2 = require('../../utils/index');

var _loaders = require('../../utils/loaders');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Morph = (0, _MeshComponent.MeshComponent)(_class = (_temp = _class2 = function (_Component) {
  (0, _inherits3.default)(Morph, _Component);

  function Morph() {
    var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    (0, _classCallCheck3.default)(this, Morph);

    var _this = (0, _possibleConstructorReturn3.default)(this, (Morph.__proto__ || (0, _getPrototypeOf2.default)(Morph)).call(this, params, Morph.defaults, Morph.instructions));

    if (params.build) {
      _this.build(params);
      (0, _get3.default)(Morph.prototype.__proto__ || (0, _getPrototypeOf2.default)(Morph.prototype), 'wrap', _this).call(_this);
    }
    return _this;
  }

  (0, _createClass3.default)(Morph, [{
    key: 'build',
    value: function build() {
      var _this2 = this;

      var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      var promise = new _promise2.default(function (resolve) {
        var Loader = params.geometry.loader;

        Loader.load(params.geometry.path, function (data, materials) {
          var material = void 0;
          if (params.material.useVertexColors) {
            material = (0, _index2.loadMaterial)(extend(params.material, {
              morphTargets: true,
              vertexColors: _three.FaceColors
            }));
          } else if (!materials || params.material.useCustomMaterial) {
            material = (0, _index2.loadMaterial)(params.material);
          } else material = new _three.MultiMaterial(materials);

          data.computeFaceNormals();
          data.computeVertexNormals();

          // Visualization.
          var mesh = new MeshNative(data, material);
          mesh.speed = params.morph.speed;
          mesh.mixer = new _three.AnimationMixer(mesh);

          mesh.mixer.clipAction(data.animations[0]).setDuration(params.morph.duration).play();

          _this2.native = mesh;

          resolve();
        });
      });

      (0, _get3.default)(Morph.prototype.__proto__ || (0, _getPrototypeOf2.default)(Morph.prototype), 'wait', this).call(this, promise);
      return promise;
    }
  }, {
    key: 'clone',
    value: function clone() {
      return new Morph({ build: false }).copy(this);
    }
  }]);
  return Morph;
}(_Component2.Component), _class2.defaults = (0, _extends3.default)({}, _Component2.Component.defaults, {
  geometry: {
    path: '',
    loader: _loaders.JSONLoader
  },

  morph: {
    duration: 1,
    speed: 100
  }
}), _class2.instructions = (0, _extends3.default)({}, _Component2.Component.instructions, {
  geometry: ['path', 'loader'],
  morph: ['duration', 'speed']
}), _temp)) || _class;

exports.Morph = Morph;
