// whs/lib/components/meshes/Parametric.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Parametric = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _get2 = require('babel-runtime/helpers/get');

var _get3 = _interopRequireDefault(_get2);

var _class, _class2, _temp;

var _three = require('three');

var _index = require('../../physics/index.js');

var _Component2 = require('../../core/Component');

var _MeshComponent = require('../../core/MeshComponent');

var _PhysicsComponent = require('../../core/PhysicsComponent');

var _SoftbodyComponent = require('../../core/SoftbodyComponent');

var _index2 = require('../../utils/index');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Parametric = (0, _SoftbodyComponent.SoftbodyComponent)(_class = (0, _PhysicsComponent.PhysicsComponent)(_class = (0, _MeshComponent.MeshComponent)(_class = (_temp = _class2 = function (_Component) {
  (0, _inherits3.default)(Parametric, _Component);

  function Parametric() {
    var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    (0, _classCallCheck3.default)(this, Parametric);

    var _this = (0, _possibleConstructorReturn3.default)(this, (Parametric.__proto__ || (0, _getPrototypeOf2.default)(Parametric)).call(this, params, Parametric.defaults, Parametric.instructions));

    if (params.build) {
      _this.build(params);
      (0, _get3.default)(Parametric.prototype.__proto__ || (0, _getPrototypeOf2.default)(Parametric.prototype), 'wrap', _this).call(_this);
    }
    return _this;
  }

  (0, _createClass3.default)(Parametric, [{
    key: 'build',
    value: function build() {
      var _this2 = this;

      var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      var material = (0, _index2.loadMaterial)(params.material);

      var MeshNative = void 0;

      if (this.physics && this.params.softbody) MeshNative = _index.SoftMesh;else if (this.physics && this.physics.type === 'concave') MeshNative = _index.ConcaveMesh;else if (this.physics) MeshNative = _index.ConvexMesh;else MeshNative = _three.Mesh;

      return new _promise2.default(function (resolve) {
        _this2.native = new MeshNative(_this2.buildGeometry(params), material, _this2.params);

        resolve();
      });
    }
  }, {
    key: 'buildGeometry',
    value: function buildGeometry() {
      var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      var GConstruct = params.buffer && !params.softbody ? _three.ParametricBufferGeometry : _three.ParametricGeometry;

      return new GConstruct(params.geometry.func, params.geometry.slices, params.geometry.stacks);
    }
  }, {
    key: 'clone',
    value: function clone() {
      return new Parametric({ build: false }).copy(this);
    }
  }, {
    key: 'g_func',
    set: function set(val) {
      this._native.geometry = this.buildGeometry(this.updateParams({ geometry: { func: val } }));
    },
    get: function get() {
      return this._native.geometry.parameters.func;
    }
  }, {
    key: 'g_slices',
    set: function set(val) {
      this._native.geometry = this.buildGeometry(this.updateParams({ geometry: { slices: val } }));
    },
    get: function get() {
      return this._native.geometry.parameters.slices;
    }
  }, {
    key: 'g_stacks',
    set: function set(val) {
      this._native.geometry = this.buildGeometry(this.updateParams({ geometry: { stacks: val } }));
    },
    get: function get() {
      return this._native.geometry.parameters.stacks;
    }
  }]);
  return Parametric;
}(_Component2.Component), _class2.defaults = (0, _extends3.default)({}, _Component2.Component.defaults, {
  geometry: {
    func: function func(u, v) {
      return new THREE.Vector3(u, v, 0);
    },
    slices: 10,
    stacks: 10
  }
}), _temp)) || _class) || _class) || _class;

exports.Parametric = Parametric;
