// whs/lib/utils/ComponentUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty = require("babel-runtime/core-js/object/define-property");

var _defineProperty2 = _interopRequireDefault(_defineProperty);

var _assign = require("babel-runtime/core-js/object/assign");

var _assign2 = _interopRequireDefault(_assign);

exports.$wrap = $wrap;
exports.$extend = $extend;
exports.$define = $define;
exports.$defaults = $defaults;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function $wrap(target) {
  var _proto = target.prototype || target.__proto__;

  return {
    onCallConstructor: function onCallConstructor(callback) {
      _proto.callConstructor = function (old) {
        return function (scope) {
          old(scope);
          callback(scope);
        };
      }(_proto.callConstructor);
    },
    onCallWrap: function onCallWrap(callback) {
      _proto.callWrap = function (old) {
        return function (scope) {
          for (var _len = arguments.length, tags = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
            tags[_key - 1] = arguments[_key];
          }

          old.apply(undefined, [scope].concat(tags));
          callback.apply(undefined, [scope].concat(tags));
        };
      }(_proto.callWrap);
    },
    onCallAddTo: function onCallAddTo(callback) {
      _proto.callAddTo = function (old) {
        return function (scope) {
          old(scope);
          callback(scope);
        };
      }(_proto.callAddTo);
    },
    onCallCopy: function onCallCopy(callback) {
      _proto.callCopy = function (old) {
        return function (scope) {
          old(scope);
          callback(scope);
        };
      }(_proto.callCopy);
    }
  };
}

function $extend(target, obj) {
  var _proto = target.prototype || target.__proto__;
  (0, _assign2.default)(_proto, obj);
}

function $define(target, obj) {
  for (var key in obj) {
    var value = obj[key];
    var _proto = target.prototype || target.__proto__;

    value.configurable = true;

    (0, _defineProperty2.default)(_proto, key, value);
  }
}

function $defaults(target, obj) {
  var defs = target.defaults || {};
  target.defaults = (0, _assign2.default)(defs, obj);
  target.params = (0, _assign2.default)(defs, obj);
}
