// whs/lib/utils/loaders.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MaterialLoader = exports.ImageLoader = exports.AudioLoader = exports.XHRLoader = exports.FontLoader = exports.TextureLoader = exports.ObjectLoader = exports.BufferGeometryLoader = exports.JSONLoader = undefined;

var _three = require('three');

// Models.
var JSONLoader = new _three.JSONLoader();
// const OBJLoader = new OBJLoaderNative();
// const BabylonLoader = new BabylonLoaderNative();
var BufferGeometryLoader = new _three.BufferGeometryLoader();
// const ColladaLoader = new ColladaLoaderNative();
// const glTFLoader = new glTFLoaderNative();
var ObjectLoader = new _three.ObjectLoader();
// const PDBLoader = new PDBLoaderNative();
// const SVGLoader = new SVGLoaderNative();
// const TGALoader = new TGALoaderNative();


// Other.
var TextureLoader = new _three.TextureLoader();
var FontLoader = new _three.FontLoader();
var XHRLoader = new _three.XHRLoader();
var AudioLoader = new _three.AudioLoader();
var ImageLoader = new _three.ImageLoader();
var MaterialLoader = new _three.MaterialLoader();
// const MTLLoader = new MTLLoaderNative();

exports.JSONLoader = JSONLoader;
exports.BufferGeometryLoader = BufferGeometryLoader;
exports.ObjectLoader = ObjectLoader;
exports.TextureLoader = TextureLoader;
exports.FontLoader = FontLoader;
exports.XHRLoader = XHRLoader;
exports.AudioLoader = AudioLoader;
exports.ImageLoader = ImageLoader;
exports.MaterialLoader = MaterialLoader;
