// whs/lib/components/cameras/OrthographicCamera.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OrthographicCamera = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _get2 = require('babel-runtime/helpers/get');

var _get3 = _interopRequireDefault(_get2);

var _class, _class2, _temp;

var _three = require('three');

var _Component2 = require('../../core/Component');

var _CameraComponent = require('../../core/CameraComponent');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var OrthographicCamera = (0, _CameraComponent.CameraComponent)(_class = (_temp = _class2 = function (_Component) {
  (0, _inherits3.default)(OrthographicCamera, _Component);

  function OrthographicCamera() {
    var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    (0, _classCallCheck3.default)(this, OrthographicCamera);

    var _this = (0, _possibleConstructorReturn3.default)(this, (OrthographicCamera.__proto__ || (0, _getPrototypeOf2.default)(OrthographicCamera)).call(this, params, OrthographicCamera.defaults, OrthographicCamera.instructions));

    _this.build(params);
    (0, _get3.default)(OrthographicCamera.prototype.__proto__ || (0, _getPrototypeOf2.default)(OrthographicCamera.prototype), 'wrap', _this).call(_this);
    return _this;
  }

  (0, _createClass3.default)(OrthographicCamera, [{
    key: 'build',
    value: function build() {
      var _this2 = this;

      var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      return new _promise2.default(function (resolve) {
        _this2.native = new _three.OrthographicCamera(params.camera.left, params.camera.right, params.camera.top, params.camera.bottom, params.camera.near, params.camera.far);

        resolve();
      });
    }
  }]);
  return OrthographicCamera;
}(_Component2.Component), _class2.defaults = (0, _extends3.default)({}, _Component2.Component.defaults, {

  camera: {
    near: 1,
    far: 1000,
    left: window.innerWidth / -2,
    right: window.innerWidth / 2,
    top: window.innerHeight / 2,
    bottom: window.innerHeight / -2
  }
}), _temp)) || _class;

exports.OrthographicCamera = OrthographicCamera;
